test/e2e/operator_test.go

```diff
@@ -64,3 +68,108 @@ func TestDefaultClusterDNSExists(t *testing.T) {
 		t.Errorf("failed to get default ClusterDNS: %v", err)
 	}
 }
+
+func TestVersionReporting(t *testing.T) {
+	var deployment *appsv1.Deployment
+	err := wait.PollImmediate(1*time.Second, 10*time.Minute, func() (bool, error) {
+		deployment = &appsv1.Deployment{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Deployment",
+				APIVersion: "apps/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "dns-operator",
+				Namespace: "openshift-dns-operator",
+			},
+		}
+		if err := sdk.Get(deployment); err != nil {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Errorf("failed to get deployment %s/%s: %v", deployment.Namespace, deployment.Name, err)
+	}
+
+	patch := []byte(`{"spec": {"template": {"spec": {"containers": [{"name":"dns-operator","env":[{"name":"RELEASE_VERSION","value":"0.0.1-test"}]}]}}}}`)
+	if err := sdk.Patch(deployment, types.StrategicMergePatchType, patch); err != nil {
+		t.Fatalf("failed to patch dns operator to new version: %v", err)
+	}
+
+	err = wait.PollImmediate(1*time.Second, 3*time.Minute, func() (bool, error) {
+		co := &configv1.ClusterOperator{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "ClusterOperator",
+				APIVersion: "config.openshift.io/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "dns",
+			},
+		}
+		if err := sdk.Get(co); err != nil {
+			return false, nil
+		}
+
+		for _, v := range co.Status.Versions {
+			if v.Name == "operator" && v.Version == "0.0.1-test" {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to observe updated version reported in dns clusteroperator status: %v", err)
+	}
+}
+
+func TestCoreDNSImageUpgrade(t *testing.T) {
```

Comment:
Although I think this probably works for the intended case, I'm afraid it's possible that it _can_ break the deployment for any future subsequent tests because the `openshift/origin-coredns:latest` may not be valid in all test contexts (plus who knows what image we'll actually get). If the only assertion is that the deployment renders the new image, maybe a unit test would make more sense?

Reply:
@Miciah had the idea to restore the original image in a defer to avoid having to refactor for unit testing.

Reply:
https://github.com/openshift/cluster-dns-operator/pull/83